Jgalp 128 publish fast slam visualization information to rviz

juliagranato opened the pull request.

juliagranato committed:
JGALP-123: Publishes information needed to visualize robot path

To be tested on the robot tomorrow; TODO:use conditional compile for visualization info

juliagranato committed:
JGALP-128: Modifies visuzalize_landmarks()-mAdded a frame_id parameter to visualize_landmarks() so that the function can be reused to visualize the landmarks in the particle filter

juliagranato committed:
Removed Position Calibration

reset_pose service call added to launch file so there is no need to subtract the initial pose; conditional compile for path visualization added; this has been tested in RViz and it works!

WillWu88 commented on src/create3-fastslam.cpp:
```diff
@@ -21,6 +21,8 @@
 #include "lidar-subscriber.h"
 using namespace std::chrono_literals;
 
+#define VISUALIZE_ROB_PATH
```
Why is this being separately defined here if it's already an option? I think the reason why your compile option is not triggering is because you are not linking it to the executable with `target_compile_definitions`.

WillWu88 replied:
Check line 50 of the `CMakeLists.txt` file

juliagranato replied:
I figured out that it wouldn't work before because I had it as an #elifdef after #ifdef USE_ROB_ODOM. I will double check that its only defined in one place

WillWu88 commented on src/create3-fastslam.cpp:
```diff
@@ -46,39 +48,33 @@ class FastSLAMC3: public rclcpp::Node
 public:
   FastSLAMC3(): Node("fastslam_create3"){
-
+  rclcpp::QoS subscriber_qos = rclcpp::QoS(10);
```
I would be very careful here about the life span of the qos profile. Before the object is managed by a smart pointer, so the profile object will be alive as long as an instance of this class is alive. Here you are creating a local object whose scope is confined in the constructor, so it gets destroyed after the constructor finishes executing. Unless you are passing in by value for `create_subscription`, which I am not sure and you should double check, the `subscriber_qos` profile will go out of scope and this might create problems for the subscriber since it will be alive longer. How we did it before is the safe way but double check to see if your way can also work.

juliagranato replied:
It definitely worked when I tested it out on the robot today, because I am passing by value into create subscription. I can make it a smart pointer if you'd prefer

WillWu88 replied:
Sound good! Just want to make sure. Either way is fine but smart pointer would be nice.

WillWu88 commented on src/create3-fastslam.cpp:
```diff
@@ -46,39 +48,33 @@ class FastSLAMC3: public rclcpp::Node
 public:
   FastSLAMC3(): Node("fastslam_create3"){
-
+  rclcpp::QoS subscriber_qos = rclcpp::QoS(10);
```
Please fix the indent level of the constructor body

WillWu88 commented on src/create3-fastslam.cpp:
```diff
@@ -96,73 +92,68 @@ class FastSLAMC3: public rclcpp::Node
 
   //robot parameters
-  struct Pose2D m_rob_pose_delta;
-  bool m_pose_calibrated; // discard first sample
```
Did you eventually figure out how to reset the odometry through this node? I guess we can manually reset it every time before we run this but it would be nice if there is an automatic way of resetting pose, whether through a separate process or maybe through the launch file.

juliagranato replied:
I added it to the launch file, so we no longer have to worry about it

WillWu88 replied:
Nice!

WillWu88 commented on src/wheel_tick_odometry.cpp:
Overall looks good. Please address the comments and be mindful of the neatness of the code. Let me know how the robot test go. It would be cool to plot a raw robot odometry vs ground truth through RVIZ with this new addition. Good work

juliagranato replied:
Emma is working on visualizing vicon, we'll going to look into the Vicon bridge. I tested this code and it does accurately visualize the robot path. Vim indents differently than your text editor so sometimes it gets messed up, I'll go back and fix the indentation

WillWu88 replied:
Thanks!

WillWu88 replied:
Perhaps this motivate us to include an automatic formatter in our repo. I will make a ticket

juliagranato committed:
JGALP-128: Minor fixes suggested in last PR

fixed constructor indentation, correctly added VISUALIZE_ROB_PATH to CMakeLists.txt; made QoS a shared pointer

juliagranato committed:
JGALP-128: Efficiently visualizes robot path and landmark observations

-minor fixes suggested in PR on 8/6/24:
        -fixed indentation in constructor
        -modified CMakeLists.txt to properly incorporate VISUALIZE_ROB_PATH
         and VISUALIZE_LANDMARK_OBSERVATIONS
        -Made QoS Profile a shared pointer
-added doxygen style commenting to lidar-subscriber.h
-overloaded visualize_landmarks to be compatible with FastSLAMPF
-overloaded laserscan_to_landmarks so that landmark_observation_publisher parameter is only needed when
        the VISUALIZE_LANDMARK_OBSERVATIONS flag is on
-added conditional compile blocks to create3-fastlslam.cpp in accordance with the above changes
cd '/home/willwu/slam_ros_ws/src/fastslam_ros'

WillWu88 commented on include/lidar-subscriber.h:
```diff
@@ -7,13 +12,30 @@
 
-std::queue<Observation2D> laserscan_to_landmarks(const sensor_msgs::msg::LaserScan::SharedPtr msg, rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr observation_visualization_publisher);
+/**
+ * @overload std::queue<Observation2D> laserscan_to_landmarks(const sensor_msgs::msg::LaserScan::SharedPtr msg);
```
Nice use of docstring here!

WillWu88 commented on src/create3-fastslam.cpp:
```diff
@@ -95,74 +89,71 @@ class FastSLAMC3: public rclcpp::Node
 
 void FastSLAMC3::lm_callback(const sensor_msgs::msg::LaserScan::SharedPtr msg){
+#ifdef VISUALIZE_LANDMARK_OBSERVATIONS
   std::queue<Observation2D> lidar_landmarks = laserscan_to_landmarks(msg, m_observation_visualization_pub);
```
Minor: I think this function here has enough of a difference to warrant a different name than `laserscan_to_landmarks`, but we can keep it for now.

WillWu88 commented on src/lidar_subscriber.cpp:
```diff
@@ -46,6 +46,73 @@ static void merge_landmarks(Observation2D &first_landmark, Observation2D &last_l
+}
+
+#ifdef VISUALIZE_LANDMARK_OBSERVATIONS
 void visualize_landmarks(std::queue<Observation2D> landmarks, rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr observation_visualization_publisher){
```
This is really confusing and should not be overloaded. The previous instance is visualizing raw point cloud to my understanding, and this is visualizing the centroids of observed landmarks. I think this warrants a different name here. This function also doesn't need to be wrapped in conditional compile.

Remember to change the header file!

juliagranato replied:
I can rename the functions to be more clear. I figured that since we only need this version if we wanted to test laserscan_to_landmarks so I thought it was more efficient to wrap it in a conditional compile since we won't use it most of the time

WillWu88 commented on src/lidar_subscriber.cpp:
```diff
@@ -46,6 +46,73 @@ static void merge_landmarks(Observation2D &first_landmark, Observation2D &last_l
+  message.scale.x = 0.04;
+  message.scale.y = 0.04;
+  auto color = std_msgs::msg::ColorRGBA();
+  color.r = 255;
```
These hard constants should be eliminated in the code (including below on line 129) and stored somewhere else (probably in the header file). If we reach a critical mass of configuration options we might want to start looking into json files as an option to store settings.

WillWu88 reviewed (commented):
Looks good. Just some minor function signature/naming changes and we should be good to go.

juliagranato committed:
JGALP-128: LM observation and visualization clean up

-use two count int instead of a queue as suggested in ALP-137
-renamed functions to reduce confusion and overloading as suggested in review
-renamed files to be more clear
-changed function names to camelCase
-added constants to header file rather than hard coding values
-fixed some improper declarations in create3-fastslam.cpp
-fixed implementation of the VISUALIZE_LANDMARK_OBSERVATIONS compile flag

juliagranato commented on src/create3-fastslam.cpp:
Note that path visualization has been tested on the robot, it works!

WillWu88 replied:
Good. Merging.

WillWu88 commented on CMakeLists.txt:
```diff
@@ -27,32 +27,33 @@ find_package(geometry_msgs REQUIRED)
   create3
-  LidarSubscriberLib
+  LMLib
   DESTINATION lib/${PROJECT_NAME}
```
I don't think this is necessary here, since `LMLib` isn't a ROS executable. I missed it earlier somehow.

WillWu88 committed:
Merge branch 'main' of

https://github.com/USC-ACTLab/fastslam_ros into JGALP-128-publish-fast-slam-visualization-information-to-rviz
